File: mcp-server/src/routes/coverage.test.ts
import { describe, it, expect } from "vitest";
import { parseLcov, parseCobertura } from "./coverage.js";

describe("parseLcov", () => {
  it("returns a length-1 range for one covered line", () => {
    const lcov = "SF:src/runner.ts\nDA:42,1\nend_of_record\n";
    expect(parseLcov(lcov)).toEqual([
      { file: "src/runner.ts", startLine: 42, endLine: 42 },
    ]);
  });

  it("collapses three contiguous covered lines into one range", () => {
    const lcov = "SF:src/runner.ts\nDA:10,1\nDA:11,1\nDA:12,1\nend_of_record\n";
    expect(parseLcov(lcov)).toEqual([
      { file: "src/runner.ts", startLine: 10, endLine: 12 },
    ]);
  });

  it("attributes each chunk to its own SF record across two files", () => {
    const lcov =
      "SF:src/a.ts\nDA:1,1\nend_of_record\nSF:src/b.ts\nDA:5,1\nend_of_record\n";
    expect(parseLcov(lcov)).toEqual([
      { file: "src/a.ts", startLine: 1, endLine: 1 },
      { file: "src/b.ts", startLine: 5, endLine: 5 },
    ]);
  });
});

describe("parseCobertura", () => {
  it("collapses contiguous covered lines per class into ranges", () => {
    const xml = `<coverage>
 <packages><package><classes>
  <class filename="src/a.ts"><lines>
   <line number="10" hits="3"/>
   <line number="11" hits="1"/>
   <line number="20" hits="2"/>
  </lines></class>
 </classes></package></packages>
</coverage>`;
    expect(parseCobertura(xml)).toEqual([
      { file: "src/a.ts", startLine: 10, endLine: 11 },
      { file: "src/a.ts", startLine: 20, endLine: 20 },
    ]);
  });
});

File: apps/lore-code-trace/coverage_test.go
package main

import (
	"reflect"
	"testing"
)

func TestParseLcovCoverageKeepsHitLinesAndCollapsesRanges(t *testing.T) {
	lcov := "TN:t\nSF:src/a.go\nDA:1,1\nDA:2,1\nDA:3,0\nDA:5,2\nend_of_record\n"
	got := parseLcovCoverage(lcov)
	want := []CoveredChunk{
		{File: "src/a.go", StartLine: 1, EndLine: 2},
		{File: "src/a.go", StartLine: 5, EndLine: 5},
	}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("got %+v, want %+v", got, want)
	}
}

func TestParseCoberturaCoverageKeepsHitLines(t *testing.T) {
	xml := `<class filename="src/b.go"><lines>` +
		`<line number="1" hits="1"/><line number="2" hits="0"/><line number="3" hits="4"/>` +
		`</lines></class>`
	got := parseCoberturaCoverage(xml)
	want := []CoveredChunk{
		{File: "src/b.go", StartLine: 1, EndLine: 1},
		{File: "src/b.go", StartLine: 3, EndLine: 3},
	}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("got %+v, want %+v", got, want)
	}
}

func TestCollapseRangesSortsAndMergesContiguous(t *testing.T) {
	got := collapseRanges("f", []int{3, 1, 2, 7})
	want := []CoveredChunk{
		{File: "f", StartLine: 1, EndLine: 3},
		{File: "f", StartLine: 7, EndLine: 7},
	}
	if !reflect.DeepEqual(got, want) {
		t.Errorf("got %+v, want %+v", got, want)
	}
}
